File: bfs & dfs/apartment.py
#백준 2667 단지번호 붙이기

n = int(input())
graph = []
home_cnt = []
cnt = 0
result = 0

for i in range(n):
    graph.append(list(map(int, input())))


def dfs(x, y):
    if x < 0 or x >= n or y < 0 or y >= n:
        return False
  
    if graph[x][y] == 1:
        graph[x][y] = 0
        global cnt
        cnt += 1
        dfs(x-1, y)
        dfs(x+1, y)
        dfs(x, y-1)
        dfs(x, y+1)
        return True
    
    return False


for i in range(n):
    for j in range(n):
        if dfs(i, j) != 0:
            result += 1
            home_cnt.append(cnt)
            cnt = 0

print(result)
home_cnt.sort()
for i in range(result):
    print(home_cnt[i])

File: bfs & dfs/apartment.js
// 백준 2667번 단지 번호 붙이기

const fs = require("fs");
const input = fs.readFileSync("input.txt").toString().trim().split("\n");

const n = Number(input.shift());
const apartment = [];

for (let i = 0; i < n; i++) {
  apartment.push(input[i].split("").map(Number));
}

const dfs = (x, y) => {
  if (x <= -1 || x >= n || y <= -1 || y >= n) return 0;
  if (apartment[x][y] >= 1) {
    apartment[x][y] = -1; // 방문 처리
    let result = 1;
    result += dfs(x - 1, y);
    result += dfs(x + 1, y);
    result += dfs(x, y + 1);
    result += dfs(x, y - 1);
    return result;
  }
  return 0;
};

let ans = [];

for (let i = 0; i < n; i++) {
  for (let j = 0; j < n; j++) {
    let current = dfs(i, j);
    if (current > 0) ans.push(current);
  }
}

ans.sort((a, b) => a - b);
console.log(ans.length + "\n" + ans.join("\n"));
